File: tba/src/main/java/de/gaalop/productComputer2/PermutableIterator.java
package de.gaalop.productComputer2;

import java.util.Arrays;

/**
 *
 * @author Christian Steinmetz
 */
public class PermutableIterator implements Permutable {

    private IntArray cur;
    private int curNo;
    private int count;

    private int[] lengths;

    public void initialize(int[] lengths) {
        this.lengths = lengths;
        Integer[] array = new Integer[lengths.length];
        Arrays.fill(array, 0);
        curNo = 0;
        count = 1;
        for (int i=0;i<lengths.length;i++)
            count *= lengths[i];
        cur = new IntArray(array);
    }

    public IntArray getNextPermutation() {
        IntArray result = cur;
        curNo++;
        if (!hasNextPermutation()) return result;
        //Integer[] curNew = Arrays.copyOf(cur.getArray(), cur.getArray().length);
        Integer[] arr = cur.getArray();
        Integer[] curNew = new Integer[arr.length];
        for (int i=0;i<curNew.length;i++)
            curNew[i] = new Integer(arr[i]);

        
        
        int pos = curNew.length;
        boolean ueberlauf = true;
        while (ueberlauf) {
            pos--;
            ueberlauf = (curNew[pos]+1 == lengths[pos]);
            if (ueberlauf) 
                curNew[pos] = 0;
            else
                curNew[pos]++;
        }

        cur = new IntArray(curNew);
        return result;
    }

    public boolean hasNextPermutation() {
        return (curNo<count);
    }

}

File: productComputerC/PermutableIterator.cpp
/*
 * PermutableIterator.cpp
 *
 *  Created on: 22.11.2011
 *      Author: christian
 */

#include "PermutableIterator.h"

PermutableIterator::PermutableIterator() {
}

PermutableIterator::~PermutableIterator() {
}

void PermutableIterator::initialize(const intVector& lengths, intVector& permutation) {
	this->lengths = lengths;
	permutation.reserve(lengths.size());
	curNo = 0;
	count = 1;
	for (unsigned int i=0;i<lengths.size();++i) {
		count *= lengths[i];
		permutation.push_back(0);
	}
}

void PermutableIterator::getNextPermutation(intVector& permutation) {
	curNo++;
	if (!(curNo<count)) return;

	int pos = permutation.size();
	bool ueberlauf = true;
	while (ueberlauf) {
		pos--;
		ueberlauf = (permutation[pos]+1 == lengths[pos]);
		if (ueberlauf)
			permutation[pos] = 0;
		else
			permutation[pos]++;
	}
}
